addons/website_profile/controllers/main.py

```diff
@@ -0,0 +1,128 @@
+    @http.route('/profile/edit', type='http', auth="user", website=True)
+    def view_user_profile_edition(self, **kwargs):
+        countries = request.env['res.country'].search([])
+        values = self._prepare_user_values(searches=kwargs)
+        values.update({
+            'email_required': kwargs.get('email_required'),
+            'countries': countries,
+            'notifications': self._get_notifications(),
+        })
+        return request.render("website_profile.user_profile_edit_main", values)
+
+    def _prepare_save_edited_profile_values(self, user, **kwargs):
+        values = {
+            'name': kwargs.get('name'),
+            'website': kwargs.get('website'),
+            'email': kwargs.get('email'),
+            'city': kwargs.get('city'),
+            'country_id': int(kwargs.get('country')) if kwargs.get('country') else False,
+            'website_description': kwargs.get('description'),
+        }
+
+        if 'clear_image' in kwargs:
+            values['image'] = False
+        elif kwargs.get('ufile'):
+            image = kwargs.get('ufile').read()
+            values['image'] = base64.b64encode(image)
+
+        if request.uid == user.id:  # the controller allows to edit only its own privacy settings; use partner management for other cases
+            values['website_published'] = kwargs.get('website_published') == 'True'
+        return values
+
+    @http.route('/profile/user/<model("res.users"):user>/save', type='http', auth="user", methods=['POST'], website=True)
+    def save_edited_profile(self, user, **kwargs):
+        values = self._prepare_save_edited_profile_values(user, **kwargs)
+        user.write(values)
+        return werkzeug.utils.redirect("/profile/user/%d" % user.id)
+
+    # Ranks
+    # ---------------------------------------------------
+    @http.route('/profile/ranks', type='http', auth="public", website=True)
+    def ranks(self, **kwargs):
+        Rank = request.env['gamification.karma.rank']
+        ranks = Rank.sudo().search([])
+        ranks = sorted(ranks, key=lambda b: b.karma_min)
+        values = {
+            'ranks': ranks,
+        }
+        return request.render("website_profile.rank_main", values)
+
+    # Badges
+    # ---------------------------------------------------
+    def _get_notifications(self):
+        badge_subtype = request.env.ref('gamification.mt_badge_granted')
+        if badge_subtype:
+            msg = request.env['mail.message'].search([('subtype_id', '=', badge_subtype.id), ('needaction', '=', True)])
+        else:
+            msg = list()
+        return msg
+
+    def _prepare_badges_domain(self, **kwargs):
+        """
+        Hook for other modules to restrict the badges showed on profile page, depending of the context
+        """
+        domain = [('website_published', '=', True)]
+        if 'category' in kwargs:
+            domain.append(('challenge_ids.category', '=', kwargs.get('category')))
```

Comment:
When having an helper method to return domains, try to use only normalized domain.
So, you will need to use `from odoo.osv import expression` and `expression.AND() / expression.OR` ;)